File: packages/openmemory-py/src/openmemory/core/vector/s3.py
import logging
import asyncio
import json
from typing import List, Optional, Dict, Any
import numpy as np

from ..vector_store import VectorStore, VectorRow

logger = logging.getLogger("vector_store.s3")

class S3VectorStore(VectorStore):
    def __init__(self, bucket: str, prefix: str = "om-vectors/"):
        self.bucket = bucket
        # S3 Vectors uses indices within vector buckets, not prefixes! Let's treat prefix as the index name.
        self.index_name = prefix.strip('/') 
        self._s3v_client = None
        self._initialized = False

    def _get_client(self):
        """Lazy load boto3 client."""
        if self._s3v_client is None:
            import boto3
            from ..config import env
            import os
            
            # Extract region from config or env, defaulting to us-east-1
            region = env.aws_region or os.getenv("AWS_DEFAULT_REGION", "us-east-1")
            
            # Use default session which automatically inherits ECS IAM roles and environment variables
            # Use s3vectors for Amazon S3 Vector functionality
            self._s3v_client = boto3.client('s3vectors', region_name=region)
        return self._s3v_client

    async def _ensure_index_exists(self, dim: int):
        if self._initialized:
             return
             
        s3v = self._get_client()
        
        def _check_and_create():
            try:
                # First check if the index already exists.
                # Currently s3vectors lacks a describe_index that doesn't throw if absent,
                # but we can try listing or just attempt to create and catch ResourceInUseException (or similar)
                import botocore.exceptions
                try:
                    s3v.create_index(
                        vectorBucketName=self.bucket,
                        indexName=self.index_name,
                        dimension=dim,
                        dataType='float32',
                        distanceMetric='cosine' # OpenMemory standard
                    )
                    logger.info(f"Created new S3 Vector index '{self.index_name}' in bucket '{self.bucket}'")
                except botocore.exceptions.ClientError as e:
                    error_code = e.response['Error']['Code']
                    if error_code in ('ConflictException', 'ResourceInUseException', 'BucketAlreadyOwnedByYou'):
                        # Index likely exists already.
                        pass
                    else:
                        logger.warning(f"Error checking/creating S3 Vector index: {e}")
            except Exception as e:
                logger.warning(f"Error handling S3 vector index creation: {e}")

        await asyncio.to_thread(_check_and_create)
        self._initialized = True

    async def storeVector(self, id: str, sector: str, vector: List[float], dim: int, user_id: Optional[str] = None):
        await self._ensure_index_exists(dim)
        s3v = self._get_client()
        
        # Amazon S3 Vectors expects vectors alongside metadata.
        # We store the sector and user_id in the metadata object.
        metadata = {
            "sector": sector,
            "user_id": user_id or ""
        }

        # The s3vectors PutVectors API expects a list of objects
        def _upload():
            s3v.put_vectors(
                vectorBucketName=self.bucket,
                indexName=self.index_name,
                vectors=[
                    {
                        "key": id,
                        "data": {"float32": vector},
                        "metadata": metadata
                    }
                ]
            )

        await asyncio.to_thread(_upload)

    async def getVectorsById(self, id: str) -> List[VectorRow]:
        s3v = self._get_client()

        def _fetch():
            try:
                resp = s3v.get_vectors(
                    vectorBucketName=self.bucket,
                    indexName=self.index_name,
                    keys=[id],
                    returnData=True,
                    returnMetadata=True
                )
                return resp.get('vectors', [])
            except Exception as e:
                logger.warning(f"Could not get vector {id} from S3 Vectors: {e}")
                return []

        vectors_data = await asyncio.to_thread(_fetch)
        
        res = []
        for v in vectors_data:
            meta = v.get('metadata', {})
            
            # The data returned matches the struct injected, e.g. {"float32": [...]}
            vec_data = v.get("data", {})
            vector_values = vec_data.get("float32", [])
            
            res.append(VectorRow(
                id=v["key"],
                sector=meta.get("sector", ""),
                vector=vector_values,
                dim=len(vector_values)
            ))
        return res

    async def getVector(self, id: str, sector: str) -> Optional[VectorRow]:
        rows = await self.getVectorsById(id)
        for r in rows:
            if r.sector == sector:
                return r
        return None

    async def deleteVectors(self, id: str):
        s3v = self._get_client()

        def _delete():
            try:
                s3v.delete_vectors(
                    vectorBucketName=self.bucket,
                    indexName=self.index_name,
                    keys=[id]
                )
            except Exception as e:
                logger.warning(f"Failed to delete {id} from S3 Vectors: {e}")

        await asyncio.to_thread(_delete)

    async def search(self, vector: List[float], sector: str, k: int, filter: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        s3v = self._get_client()

        # To avoid strict and undocumented S3 Vector metadata schema filtering requirements
        # at index creation time, we will over-fetch and filter locally.
        fetch_k = k * 5 if (sector or filter) else k

        def _query():
            kwargs = {
                "vectorBucketName": self.bucket,
                "indexName": self.index_name,
                "queryVector": {"float32": vector},
                "topK": fetch_k,
                "returnMetadata": True,
                "returnDistance": True
            }
            try:
                resp = s3v.query_vectors(**kwargs)
                return resp.get('vectors', [])
            except Exception as e:
                logger.error(f"Failed to search S3 Vectors: {e}")
                return []

        matches = await asyncio.to_thread(_query)
        
        results = []
        for match in matches:
            meta = match.get('metadata', {})
            
            # Local filtering
            if sector and meta.get('sector') != sector:
                continue
            if filter and filter.get('user_id') and meta.get('user_id') != filter['user_id']:
                continue
                
            results.append({
                "id": match["key"],
                "similarity": 1.0 - match.get("distance", 0.0) # Convert distance to similarity
            })
            
            if len(results) >= k:
                break

        # Ensure sorted
        results.sort(key=lambda x: x["similarity"], reverse=True)
        return results

File: packages/openmemory-js/src/core/vector/s3.ts
import { VectorStore } from "../vector_store";
import { S3VectorsClient, PutVectorsCommand, QueryVectorsCommand, GetVectorsCommand, DeleteVectorsCommand, CreateIndexCommand } from "@aws-sdk/client-s3vectors";
import { env } from "../cfg";

export class S3VectorStore implements VectorStore {
    private client: S3VectorsClient;
    private initialized = false;

    constructor() {
        this.client = new S3VectorsClient({
            region: env.AWS_REGION || "us-east-1",
            credentials: (env.AWS_ACCESS_KEY_ID && env.AWS_SECRET_ACCESS_KEY)
                ? {
                    accessKeyId: env.AWS_ACCESS_KEY_ID,
                    secretAccessKey: env.AWS_SECRET_ACCESS_KEY,
                    ...(env.AWS_SESSION_TOKEN ? { sessionToken: env.AWS_SESSION_TOKEN } : {})
                }
                : undefined
        });
    }

    private async ensureIndexExists(dim: number): Promise<void> {
        if (this.initialized) return;
        try {
            await this.client.send(new CreateIndexCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                dataType: "float32",
                dimension: dim,
                distanceMetric: "cosine"
            }));
            this.initialized = true;
        } catch (e: any) {
            if (e.name === "ResourceInUseException" || e.name === "ConflictException" || e.name === "AlreadyExistsException" || e.$metadata?.httpStatusCode === 409 || e.toString().includes("already exists") || (e.name === "BadRequestException" && e.message.includes("already exists"))) {
                this.initialized = true;
            } else {
                console.warn("[S3Vectors] Could not pre-create index (it might already exist):", e.message);
                this.initialized = true;
            }
        }
    }

    async storeVector(id: string, sector: string, vector: number[], dim: number, user_id?: string, project_id?: string): Promise<void> {
        console.error(`[S3Vectors] Storing vector 1 ${id}:`, id, sector, vector, dim, user_id, project_id);
        await this.ensureIndexExists(dim);
        console.error(`[S3Vectors] Storing vector 2 ${id}:`, id, sector, vector, dim, user_id, project_id);
        try {
            const metadata: any = {
                sector: sector,
                dim: dim.toString()
            };
            if (user_id) metadata.user_id = user_id;
            if (project_id) metadata.project_id = project_id;

            await this.client.send(new PutVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                vectors: [{
                    key: `${id}_${sector}`,
                    data: { float32: vector },
                    metadata: metadata
                }]
            }));
        } catch (e: any) {
            console.error(`[S3Vectors] Failed to store vector ${id}_${sector}:`, e?.message || e);
            throw e;
        }
    }

    async deleteVector(id: string, sector: string): Promise<void> {
        try {
            await this.client.send(new DeleteVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                keys: [`${id}_${sector}`, id]
            }));
        } catch (e) {
            console.error(`[S3Vectors] Failed to delete vector ${id}_${sector}`, e);
        }
    }

    async deleteVectors(id: string): Promise<void> {
        try {
            const keysToDelete = ["episodic", "semantic", "procedural", "emotional", "reflective"].map(s => `${id}_${s}`);
            keysToDelete.push(id);
            await this.client.send(new DeleteVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                keys: keysToDelete
            }));
        } catch (e) {
            console.error("[S3Vectors] Failed to delete vectors", id, e);
        }
    }

    async searchSimilar(sector: string, queryVec: number[], topK: number, user_id?: string, project_id?: string): Promise<Array<{ id: string; score: number }>> {
        const needsPostFilter = !!(sector || user_id || project_id);
        const fetchK = Math.min(100, needsPostFilter ? topK * 5 : topK);

        try {
            const resp = await this.client.send(new QueryVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                queryVector: { float32: queryVec },
                topK: fetchK,
                returnMetadata: true,
                returnDistance: true
            }));

            const matches = resp.vectors || [];
            const results: Array<{ id: string; score: number }> = [];

            for (const match of matches) {
                const meta = (match.metadata as Record<string, any>) || {};

                if (sector && meta.sector !== sector) continue;
                if (user_id && meta.user_id && meta.user_id !== user_id) continue;
                if (project_id && meta.project_id && meta.project_id !== project_id) continue;

                results.push({
                    id: match.key as string,
                    score: 1.0 - (match.distance || 0)
                });

                if (results.length >= topK) break;
            }

            return results.sort((a, b) => b.score - a.score);
        } catch (e) {
            console.error("[S3Vectors] Search failed:", e);
            return [];
        }
    }

    async getVector(id: string, sector: string): Promise<{ vector: number[]; dim: number } | null> {
        try {
            const resp = await this.client.send(new GetVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                keys: [`${id}_${sector}`, id],
                returnData: true,
                returnMetadata: true
            }));

            const vectors = resp.vectors || [];
            if (vectors.length === 0) return null;

            // Preferred: the one with the exact key `${id}_${sector}`
            let v = vectors.find((vec: any) => vec.key === `${id}_${sector}`);
            if (!v) v = vectors[0];

            const data = v.data?.float32 || [];
            const meta = (v.metadata as Record<string, any>) || {};

            return {
                vector: data,
                dim: parseInt((meta.dim as string) || data.length.toString())
            };
        } catch (e) {
            console.warn(`[S3Vectors] Could not get vector ${id}_${sector}:`, e);
            return null;
        }
    }

    async getVectorsById(id: string): Promise<Array<{ sector: string; vector: number[]; dim: number }>> {
        try {
            const keysToFetch = ["episodic", "semantic", "procedural", "emotional", "reflective"].map(s => `${id}_${s}`);
            keysToFetch.push(id);
            const resp = await this.client.send(new GetVectorsCommand({
                vectorBucketName: env.s3_bucket,
                indexName: env.s3_index_name,
                keys: keysToFetch,
                returnData: true,
                returnMetadata: true
            }));
            const vectors = resp.vectors || [];
            return vectors.map((v: any) => ({
                sector: ((v.metadata as Record<string, any>)?.sector as string) || "default",
                vector: v.data?.float32 || [],
                dim: parseInt(((v.metadata as Record<string, any>)?.dim as string) || (v.data?.float32?.length || 0).toString())
            }));
        } catch (e) {
            return [];
        }
    }

    async getVectorsBySector(sector: string): Promise<Array<{ id: string; vector: number[]; dim: number }>> {
        // Unused by core logic largely except in legacy/migration ops
        return [];
    }
}
